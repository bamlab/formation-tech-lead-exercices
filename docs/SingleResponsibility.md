# Single Responsibility principle

## What we are trying to achieve

An app with 4 different styles of buttons.
The basic button has a few custom properties:

- icon
  - can be a name (in this case we display a letter to symplify)
  - can be a component and in this case we have a border
- disabled
- color scheme

![](./buttons.png)

## The issue

The code is not very re-usable. Everytime we want a button we have to specify the color. The disabled style has to be handled every time at the parent level.

Let's try to think about the responsibilities and how we can make it easier.
