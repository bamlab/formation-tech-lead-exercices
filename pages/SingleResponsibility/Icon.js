import React from 'react';
import {Text} from 'react-native';

const FakeIcon = ({style, letter}) => {
  return <Text style={style}>{letter}</Text>;
};

/**
 * Mock Icon component. Returns a Letter just as if it were an icon.
 * @param props style and name (graph, bolt, plus, minus)
 * @returns Element
 */
export const Icon = ({name, style}) => {
  switch (name) {
    case 'arrow':
      return <FakeIcon style={style} letter="A" />;
    case 'graph':
      return <FakeIcon style={style} letter="G" />;
    case 'bolt':
      return <FakeIcon style={style} letter="B" />;
    case 'plus':
      return <FakeIcon style={style} letter="P" />;
    default:
      throw new Error(`Icon ${name} does not exist`);
  }
};
