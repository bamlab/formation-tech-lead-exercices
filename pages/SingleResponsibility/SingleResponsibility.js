/* eslint-disable react-native/no-inline-styles */
import React from 'react';
import {View} from 'react-native';
import {Button} from './Button/Button';

const ImageIcon = () => {
  return (
    <View
      style={{
        height: 15,
        width: 15,
        backgroundColor: 'salmon',
      }}
    />
  );
};

export const SingleResponsibility = () => {
  return (
    <View
      style={{
        height: 400,
        justifyContent: 'space-around',
        padding: 30,
      }}>
      {/* basic button */}
      <Button
        icon="arrow"
        color="grey"
        backgroundColor="white"
        onPress={() => console.log(1)}
      />
      {/* disabled button */}
      <Button
        icon="graph"
        color="white"
        backgroundColor="grey"
        onPress={() => console.log(2)}
        isDisabled={true}
      />
      {/* button with an image as icon */}
      <Button
        icon={<ImageIcon />}
        color="red"
        backgroundColor="papayawhip"
        onPress={() => console.log(3)}
      />
      {/* button with a different color scheme */}
      <Button
        icon="plus"
        color="red"
        backgroundColor="papayawhip"
        onPress={() => console.log(4)}
      />
    </View>
  );
};
