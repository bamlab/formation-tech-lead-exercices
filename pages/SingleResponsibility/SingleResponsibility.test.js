import React from 'react';

import {render} from '@testing-library/react-native';
import {SingleResponsibility} from './SingleResponsibility';

it('renders correctly', () => {
  const {toJSON} = render(<SingleResponsibility />);

  expect(toJSON()).toMatchSnapshot();
});
