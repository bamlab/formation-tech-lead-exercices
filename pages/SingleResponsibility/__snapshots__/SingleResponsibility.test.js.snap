// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "height": 400,
      "justifyContent": "space-around",
      "padding": 30,
    }
  }
>
  <View
    accessibilityState={
      Object {
        "disabled": false,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    hitSlop={
      Object {
        "bottom": 20,
        "left": 20,
        "right": 20,
        "top": 20,
      }
    }
    nativeID="animatedComponent"
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "white",
        "borderRadius": 40,
        "height": 40,
        "justifyContent": "center",
        "opacity": 1,
        "width": 40,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "grey",
        }
      }
    >
      A
    </Text>
  </View>
  <View
    accessibilityState={
      Object {
        "disabled": true,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    hitSlop={
      Object {
        "bottom": 20,
        "left": 20,
        "right": 20,
        "top": 20,
      }
    }
    nativeID="animatedComponent"
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "grey",
        "borderRadius": 40,
        "height": 40,
        "justifyContent": "center",
        "opacity": 1,
        "width": 40,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "white",
        }
      }
    >
      G
    </Text>
  </View>
  <View
    accessibilityState={
      Object {
        "disabled": false,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    hitSlop={
      Object {
        "bottom": 20,
        "left": 20,
        "right": 20,
        "top": 20,
      }
    }
    nativeID="animatedComponent"
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "papayawhip",
        "borderColor": "red",
        "borderRadius": 40,
        "borderWidth": 1,
        "height": 40,
        "justifyContent": "center",
        "opacity": 1,
        "width": 40,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "salmon",
          "height": 15,
          "width": 15,
        }
      }
    />
  </View>
  <View
    accessibilityState={
      Object {
        "disabled": false,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    hitSlop={
      Object {
        "bottom": 20,
        "left": 20,
        "right": 20,
        "top": 20,
      }
    }
    nativeID="animatedComponent"
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "papayawhip",
        "borderRadius": 40,
        "height": 40,
        "justifyContent": "center",
        "opacity": 1,
        "width": 40,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "red",
        }
      }
    >
      P
    </Text>
  </View>
</View>
`;
